pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1374,13 +1388,13 @@ func (c *PodAffinityChecker) satisfiesPodsAffinityAntiAffinity(pod *v1.Pod,
 		// Check all affinity terms.
 		topologyPairsPotentialAffinityPods := predicateMeta.topologyPairsPotentialAffinityPods
 		if affinityTerms := GetPodAffinityTerms(affinity.PodAffinity); len(affinityTerms) > 0 {
-			matchExists := c.nodeMatchesAllTopologyTerms(pod, topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
+			matchExists := c.nodeMatchesAllTopologyTerms(topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
 			if !matchExists {
 				// This pod may the first pod in a series that have affinity to themselves. In order
 				// to not leave such pods in pending state forever, we check that if no other pod
 				// in the cluster matches the namespace and selector of this pod and the pod matches
 				// its own terms, then we allow the pod to pass the affinity check.
-				if !(len(topologyPairsPotentialAffinityPods.topologyPairToPods) == 0 && targetPodMatchesAffinityOfPod(pod, pod)) {
+				if !targetPodMatchesAffinityOfPod(pod, pod) {
```

Comment:
This is now incorrect. In the previous code we were checking that when there was no pod in the whole cluster that matched the affinity of the incoming pod AND the incoming pod didn't match its own affinity, we would return "NotMatch". This PR has eliminated the first part of the condition. So, if the incoming pod's affinity is not satisfied on this node, we will not return "NotMatch" when the pod matches its own affinity, regardless of the fact that there might be other pods in the cluster that match the incoming pod's affinity.

Reply:
So is that to say, currently it needs to get an "aggregated" match result in advance which represents "if there is some node matches incoming pod's affinity".

- If the "aggregated" match result is true, then just give final fit judgement depending on `matchExists`
- If the "aggregated" match result is false, and `matchExists` is also false, then proceed to check if its podAffinity matches its own labels.

Reply:
> it needs to get an "aggregated" match result in advance which represents "if there is some node matches incoming pod's affinity".

This isn't that easy to implement based on current code.

Firstly, we can't simply do a pre-compute b/c for some of nodes, predicates would have failed before going to InterPodAffinity predicate. So an arbitrary pre-compute is impractical and inefficient.

Secondly, like what current logic is, it's computed on demand. But it should be able to achieve that if `!matchExists`, hold on until it gets a signal indicating either 1) some other node passed PodAffinity check, or 2) nodes running InterPodAffinity predicate (note a: not all nodes) have all failed on PodAffinity check.

Reply:
> Secondly, like what current logic is, it's computed on demand. But it should be able to achieve that if `!matchExists`, hold on until it gets a signal indicating either 1) some other node passed PodAffinity check, or 2) nodes running InterPodAffinity predicate (note a: not all nodes) have all failed on PodAffinity check.

Current predicates framework assumes predicates can be run independently on each node. So it's pretty challenging to have above logic implemented in InterPodAffinity predicate.

Another implementation is to do a post-handling after all predicates finish. But this requires (1) InterPodAffinity predicate is the last predicate (it's the case right now) and (2) inside InterPodAffinity predicate, PodAffinity check should also be handled at last step (specifically, in satisfiesPodsAffinityAntiAffinity(), the logic of handling affinity should be moved after logic of handling antiAffinity).

A quick implementation is [here](https://gist.github.com/Huang-Wei/87e4142709dfc204f7621e8a7b525edb).

Reply:
> @bsalamat I believe case https://gist.github.com/Huang-Wei/008dae22f0bba635ed4f1ad151f9768c is what you meant.

Yes, except that the pod is schedulable on both nodes. It has affinity to itself. Since its affinity doesn't match any pods in the cluster, but itself, its affinity become a no-op and it can be scheduled on any node as long as it passes other predicates.

> Current predicates framework assumes predicates can be run independently on each node. So it's pretty challenging to have above logic implemented in InterPodAffinity predicate.

Yes. I agree. I don't have any good solution in mind at the moment.

> Another implementation is to do a post-handling after all predicates finish. But this requires (1) InterPodAffinity predicate is the last predicate (it's the case right now) and (2) inside InterPodAffinity predicate, PodAffinity check should also be handled at last step (specifically, in satisfiesPodsAffinityAntiAffinity(), the logic of handling affinity should be moved after logic of handling antiAffinity).

Yes, but that's very brittle. These assumptions may be valid in the current implementation, but they could change in the future. We shouldn't corner ourselves by making these assumptions invariant.

Reply:
Thanks @bsalamat. Those are almost the same understanding from me. Let me think more to give a more elegant solution.

BTW: regarding

> Yes, except that the pod is schedulable on both nodes.

Maybe you mis-read that case - nodeA is a fit actually :) So only nodeA is a fit.

Reply:
> Maybe you mis-read that case - nodeA is a fit actually :) So only nodeA is a fit.

Yes, I didn't notice that. You are right. NodeA is the only one.